chown cgroup to process uid in container namespace

frasertweedale opened the pull request:
Delegating cgroups to the container enables more complex workloads,
including systemd-based workloads.  The OCI runtime-spec was
recently updated to explicitly admit such delegation, through
specification of cgroup ownership semantics:

  https://github.com/opencontainers/runtime-spec/pull/1123

Pursuant to the updated OCI runtime-spec, change the ownership of
the container's cgroup directory and particular files therein, when
using cgroups v2 and when the cgroupfs is to be mounted read/write.

As a result of this change, systemd workloads can run in isolated
user namespaces on OpenShift when the sandbox's cgroupfs is mounted
read/write.

It might be possible to implement this feature in other cgroup
managers, but that work is deferred.

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -292,6 +296,32 @@ func (m *unifiedManager) Apply(pid int) error {
 		return err
 	}
+
+	goodToChown := []string{"cgroup.procs", "cgroup.subtree_control", "cgroup.threads"}
+	member := func(haystack []string, needle string) bool {
+		for _, v := range haystack {
+			if v == needle {
+				return true
+			}
+		}
+		return false
```
Perhaps it's easier to use `map[string]struct{}` here.

Or just be direct
```
if name == "cgroup.procs" || name == "cgroup.subtree_control" || name == ....
```

frasertweedale replied:
OK, I'll use a map.

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -292,6 +296,32 @@ func (m *unifiedManager) Apply(pid int) error {
+			}
+			if path == m.path || member(goodToChown, info.Name()) {
+				return os.Chown(path, m.uid, -1)
+			} else {
```
`else` is redundant here

frasertweedale replied:
I prefer if / else to be balanced.  Does it violate a formal style guide?

kolyshkin replied:
Not really, but some linters do complain, and personally I find it redundant, too -- because of `return`.

Feel free to keep though.

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -27,14 +27,18 @@ type unifiedManager struct {
 }
 
-func NewUnifiedManager(config *configs.Cgroup, path string, rootless bool) cgroups.Manager {
+func NewUnifiedManager(config *configs.Cgroup, path string, uid int, rootless bool) cgroups.Manager {
```
I'm afraid we can't change this function signature, as other software (kubernetes, cri-o, possibly cAdvisor etc.) depend on it (unfortunately).

Maybe you can add a separate method to set the uid?

kolyshkin replied:
or a `WithUID()` modifier method.

frasertweedale replied:
Thanks, I will do as you suggest.

kolyshkin commented on libcontainer/factory_linux.go:
```diff
@@ -243,7 +243,7 @@ type LinuxFactory struct {
 
 	// NewCgroupsManager returns an initialized cgroups manager for a single container.
-	NewCgroupsManager func(config *configs.Cgroup, paths map[string]string) cgroups.Manager
+	NewCgroupsManager func(config *configs.Cgroup, paths map[string]string, pid int) cgroups.Manager
```
typo (`s/pid/uid/`)?

frasertweedale replied:
yep, thanks.

kolyshkin commented:
In general this kinda makes sense, although I'm afraid systemd might overwrite that at some random moment.

This also needs test(s).

frasertweedale commented:
> In general this kinda makes sense, although I'm afraid systemd might overwrite that at some random moment.
> 
> This also needs test(s).

You can use systemd transient unit API to set a different user.  Unfortunately, the uid needs to
have a corresponding passwd entity or else systemd refuses to proceed.

I filed an issue against systemd to relax this requirement.  It was provisionally rejected:

AkihiroSuda commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -27,14 +27,18 @@ type unifiedManager struct {
 	dbus     *dbusConnManager
+	// host UID corresponding to UID 0 in container namespace;
+	// or -1 if cgroup should not be chown'd
+	uid int
```
`-1` -> `nil` , `int` -> `*int`

AkihiroSuda commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -292,6 +296,32 @@ func (m *unifiedManager) Apply(pid int) error {
 		return err
 	}
+
+	goodToChown := []string{"cgroup.procs", "cgroup.subtree_control", "cgroup.threads"}
```
Needs to be documented

AkihiroSuda reviewed (changes requested):
For security reason, the default behavior should not be changed.
We can probably have a feature gate as an annotation 

https://github.com/containers/crun/blob/main/crun.1.md#extensions-to-oci

https://github.com/opencontainers/runc/blob/master/docs/systemd.md#auxiliary-properties

frasertweedale commented:
> For security reason, the default behavior should not be changed.
> We can probably have a feature gate as an annotation
> 
> https://github.com/containers/crun/blob/main/crun.1.md#extensions-to-oci
> 
> https://github.com/opencontainers/runc/blob/master/docs/systemd.md#auxiliary-properties

User namespaces is already behind an annotation (`io.kubernetes.cri-o.userns-mode`).  And user namespaces are somewhat hobbled without the ability to manage cgroups (for the workloads that I care about anyway).  So, do we really need a separate annotation to gate this feature?  I don't mind if the answer is yes, but something to consider...

AkihiroSuda commented:
I'm talking about OCI Runtime Spec annotations, not about Kubernetes annotations.

CRI-O may set the suggest OCI Runtime Spec annotation by default at their own risk.

frasertweedale commented:
@AkihiroSuda thanks for the clarification.  Fair enough.  I'll gate it.  I'll try to make all the suggested changes next week.

frasertweedale commented:
@AkihiroSuda what would be an appropriate key/value for the annotation?  `"org.opencontainers.runc.chown-cgroup": "{true,false}"`

AkihiroSuda commented:
>  "org.opencontainers.runc.chown-cgroup": "{true,false}"

SGTM, but in the long term we should have proper JSON entry in OCI Runtime Spec https://github.com/opencontainers/runtime-spec

frasertweedale commented:
> > "org.opencontainers.runc.chown-cgroup": "{true,false}"
> 
> SGTM, but in the long term we should have proper JSON entry in OCI Runtime Spec https://github.com/opencontainers/runtime-spec

Better would be to adjust the semantics such that the container process can create scopes/slices in its own cgroup, and manage its child processes/threads.  But that is a discussion to be had over there, not here :)

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -292,6 +296,32 @@ func (m *unifiedManager) Apply(pid int) error {
+		return false
+	}
+	if m.uid >= 0 {
+		err := filepath.Walk(m.path, func(path string, info os.FileInfo, err error) error {
```
Since you're only using info.Name(), it probably make sense to switch to filepath.WalkDir. Problem is, it is only added in go 1.16. So maybe add a TODO item:
```go
// TODO: switch to filepath.WalkDir once go < 1.16 is not supported.
```

frasertweedale replied:
Yep, I originally used WalkDir then had to change for that reason.  Happy to add the comment.

frasertweedale commented:
Thanks @kolyshkin and @AkihiroSuda for your reviews.  I implemented all of your requested changes.  I still need to add test and document the feature.

frasertweedale commented:
Ping @kolyshkin and @AkihiroSuda - are you able to re-review this PR soon?

frasertweedale commented:
I published a video demonstration of using this feature to run systemd-based workloads in isolated user namespaces on OpenShift: https://www.youtube.com/watch?v=OGAVvIJwmd0

Also a blog post that follows the same procedure: https://frasertweedale.github.io/blog-redhat/posts/2021-07-22-openshift-systemd-workload-demo.html

frasertweedale commented:
@kolyshkin, @AkihiroSuda: weekly nudge.  Anything I can do to help with review or testing?

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -292,6 +304,38 @@ func (m *unifiedManager) Apply(pid int) error {
+	}
+	if m.uid != nil {
+		// TODO: when go < 1.16 is no longer supported, switch to filepath.WalkDir
+		err := filepath.Walk(m.path, func(path string, info os.FileInfo, err error) error {
```
So, are we expecting to have sub-directories here? AFAICS there can be none, or am I missing something?

frasertweedale replied:
Just the one level at this time.  But I thought it better not to couple tightly to this assumption.  Do you prefer to explicitly chown exactly the 4 files needed at this time (the cgroup dir, and the 3 files within it).

frasertweedale replied:
I changed it to explicitly mention (and chown) the dir and 3 files needed; no more `Walk`.

kolyshkin commented on libcontainer/factory_linux.go:
```diff
@@ -327,6 +339,16 @@ func (l *LinuxFactory) Load(id string) (Container, error) {
 		root:                 containerRoot,
 		created:              state.Created,
 	}
+	if state.Config.ChownCgroup {
+		uid, err := state.Config.HostRootUID()
+		if err != nil {
+			return nil, err
+		}
+		err = c.cgroupManager.SetUID(&uid)
+		if err != nil {
+			return nil, err
+		}
+	}
```
1. Why this code needs to be in `Load`? I can't think of any code paths that can use it...

2. If this is indeed required here, maybe it's better to have a function to be called from both `Load` and `Create`.

frasertweedale replied:
Latest version of patch conveys the owner UID in the `configs.Cgroup` struct.  No changes needed in `factory_linux.go` anymore.

kolyshkin commented:
Overall I think it's a great idea to have the option of giving the container user an ability to create sub-cgroups.

I'd rather call the method `Chown` than `SetUID` as the latter has a special meaning. Using a pointer to int as its argument does not make sense to me, too (storing at as a pointer, though, to distinguish between "0" and "not set", kinda does). These things are not too important, of course, and are trivial to fix.

It bothers me that this is a new public method of cgroupManager, and out of 4 implementations that we have in libcontainer, only one implements it. Leaving cgroup v1 aside, is there anything that prevents fs2 manager to have it?

In general, I think, the API provided by the cgroup manager is not adequate for this task, and this is why this looks strange. I was thinking about changing that but alas don't have a clear idea how.

If we could do the chown in `Set` instead, it'd straighten things up a lot -- instead of a special method, we'd just add a flag to container Resources (like we did for, say, SkipDevices).

frasertweedale commented:
@kolyshkin thanks for your feedback.  I'll try and redesign it to use an extra field in `Resources`, which as you point out will avoid any API changes.

frasertweedale commented:
@kolyshkin OK, now I have a better comprehension of where `Apply` and `Set` get invoked.  I feel strongly that the setting the cgroup ownership should be a behaviour of `Apply`, not `Set`.  This is because there are alternative ways - besides `runc` performing `chown(2)` on relevant files - that the cgroup ownership could be set.  In particular, you can specify an alternative cgroup ownership via the systemd transient unit API.  In fact, this would be a nice way to do it except for one aspect of systemd's behaviour related to `passwd` entities (and they are resisting changing this behaviour; see https://github.com/systemd/systemd/issues/19781).

Nevertheless, to better support such an approach if systemd should relax its checks, or to support alternative approaches, I think `Apply` rather than `Set(Resources)` is the better approach.

To avoid the API changes and confusing propagation of the UID setting, perhaps we should add the cgroup owner UID as a field of `configs.Cgroup` rather than `configs.Resources`.  For now, I will continue along that trajectory.

kolyshkin commented:
@frasertweedale I was also thinking about adding `Create`.

There are two reasons for that:

1. In many cases (including kubernetes using runc's libcontainer/cgroups) `Apply(-1)` is used as a surrogate for `Create`.
2. In many cases `Apply(-1)` is followed by `Set`, as the cgroup is initially created with no limits set.

So, perhaps we can add `Create()` with the same argument as `Set` (so it will create a cgroup and set its resources). Now, `Apply` can be renamed to e.g. `AddProcess`, and `Set` can become `Update`.

I understand this is mostly out of scope for this PR, and is a major change of the API, but this PR creates a challenge that the existing API is seemingly failing to solve.

frasertweedale commented:
Note well: the only CI job that currently executes (i.e. does not skip) the test for this feature is the Cirrus CI Fedora 34 job.

frasertweedale commented:
@kolyshkin I updated the PR.  Using a new field in the `configs.Cgroup` struct simplified the implementation a lot.  Now there are no new methods or other API changes (besides the new struct field).

I also changed the chown behaviour to directly chown the 4 files of interest, instead of using `os.Walk`.

I don't see any reason this can't be implemented for fs2.  But it is not my use case (OpenShift, which uses systemd manager), so I'd rather leave it out of this commit.

frasertweedale commented:
@kolyshkin *nudge*.  I look forward to review of updated patch, when you have the capacity.  Thanks!

frasertweedale commented:
*nudge*: periodic request for review.

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -292,6 +292,24 @@ func (m *unifiedManager) Apply(pid int) error {
+	// the cgroup for its child processes/threads.
+	//
+	// We follow systemd's lead and chown only those same files.
+	filesToChown := []string{".", "cgroup.procs", "cgroup.subtree_control", "cgroup.threads"}
```
I remembered there's a cgroup file which lists these three, but can't find it now. Maybe I remember wrong :)

frasertweedale replied:
@kolyshkin I am not aware of it.

kolyshkin replied:
This is also documented in the kernel docs: 
https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#model-of-delegation

(no need to add a link I guess, just noticing)

frasertweedale commented:
CRI-O doesn't set the annotation, but it does propagate Pod annotations to the OCI runtime.  So for the time being you have to set two annotations.  Feel free to propose the CRI-O change, but it probably doesn't make sense to do so until this is merged, in case the annotation name changes.

m-yosefpor commented:
@frasertweedale thank you for your explanation. So for now setting both annotations will work for us.

kolyshkin commented on tests/integration/cgroups.bats:
```diff
@@ -119,6 +119,59 @@ function setup() {
+	runc run -d --console-socket "$CONSOLE_SOCKET" test_cgroup_chown
+	[ "$status" -eq 0 ]
+
+	runc exec test_cgroup_chown sh -c "ls -ld /sys/fs/cgroup | awk '{print \$3}'"
```
nit: no need to escape `$` since it's inside single quotes.

frasertweedale replied:
It is inside single quotes... inside double quotes :)  So the escape is necessary.  For example:
```shell
% echo "'$0'"
'zsh'
```

jfroy replied:
You could also use `stat`. Perhaps a bit more simple than ls + awk.

```shell
stat -c "%u %g" "/sys/fs/cgroup"
[ "$output" = "0 0" ]
```

kolyshkin commented on libcontainer/specconv/spec_linux.go:
```diff
@@ -306,6 +306,18 @@ func CreateLibcontainerConfig(opts *CreateOpts) (*configs.Config, error) {
+	// Set the host UID that should own the container's cgroup.
+	// This must be performed after setupUserNamespace, so that
+	// config.HostRootUID() returns the correct result.
+	if v := spec.Annotations["org.opencontainers.runc.chown-cgroup"]; v == "true" {
```
What if crun wants to add support for this? Can we use, say, `run.oci.cgroup.chown` or something like this?

@giuseppe PTAL

frasertweedale replied:
@kolyshkin actually, `crun` already has this behaviour:

```
% jq '.linux.uidMappings, .linux.cgroupsPath, .process.user' < config.json
[
  {
    "containerID": 0,
    "hostID": 1000,
    "size": 1
  },
  {
    "containerID": 1,
    "hostID": 100000,
    "size": 65536
  }
]
"user.slice:runc:bar"
{
  "uid": 0,
  "gid": 0
}
% sudo crun --systemd-cgroup run test 
... # in separate shell
% ls -alin /sys/fs/cgroup/user.slice/test-test.scope/container |grep 1000
40139 drwxr-xr-x. 2 1000 1000 0 Sep 10 16:09 .
40141 -rw-r--r--. 1 1000 1000 0 Sep 10 16:09 cgroup.procs
40144 -rw-r--r--. 1 1000 1000 0 Sep 10 16:09 cgroup.subtree_control
40142 -rw-r--r--. 1 1000 1000 0 Sep 10 16:09 cgroup.threads
40176 -rw-r--r--. 1 1000 1000 0 Sep 10 16:09 memory.oom.group
```

giuseppe replied:
yes, crun always chowns the cgroup to root in the user namespace

frasertweedale replied:
@kolyshkin with this in mind, what do you think about removing the annotation and just having this as the default behaviour in `runc` also?

thaJeztah replied:
I'm definitely not a fan of using annotations for this; it feels like a band-aid / hack to work around missing functionality in the runtime spec. If there _would be_ an annotation, it should be one that's defined in the runtime spec (and image-spec?), which should describe how runtimes should process these annotations.

What are the consequences of setting this as default? Does it affect non-systemd-in-container situations?

thaJeztah replied:
(apologies for going a bit orthogonal here)

I understand the "systemd in a container" use-case, but recall discussions from the past where some of those configuration changes needed were _beneficial_ for that use-case, but could have a _negative_ impact on other scenarios. I also recall other configuration changes (e.g. the `CONTAINER` env-var, which so far hasn't been described in the runtime spec) are needed to make systemd work properly inside a container.
Thinking out loud here, but would it be an idea of defining (e.g.) a concept of "profiles" in the runtime spec, that (when set) instructs the runtime to make the required changes to facilitate systemd? (`profile=systemd`)

frasertweedale replied:
@thaJeztah as for non-systemd workloads, well, it means they can manage their own cgroups too.  I do not think it would break anything (never say never...)

As for changes to support systemd that break or impair other use cases, do you have examples?
In my case, this experimental version of runc, combined with the CRI-O user namespace support and worker nodes configured with cgroupsv2, is sufficient to run systemd-based container on OpenShift.

I am wrestling with some other problems with more advanced systemd features (e.g. *systemd-tmpfiles* not playing well with `emptyDir` mounts) but these can be worked around.

frasertweedale replied:
> I'm definitely not a fan of using annotations for this; it feels like a band-aid / hack to work around missing functionality in the runtime spec. If there _would be_ an annotation, it should be one that's defined in the runtime spec (and image-spec?), which should describe how runtimes should process these annotations.

I will propose a clarification to the OCI runtime spec to specify that cgroups created by the OCI runtime should be owned by the container's uid 0.  At the very least some implementation notes/advice should be adopted into the OCI documents.

frasertweedale replied:
runtime-spec PR: https://github.com/opencontainers/runtime-spec/pull/1123

thaJeztah replied:
Thanks!

> As for changes to support systemd that break or impair other use cases, do you have examples?


Don't have direct examples, and I would have to dig into history for that.


> In my case, this experimental version of runc, combined with the CRI-O user namespace support and worker nodes configured with cgroupsv2, is sufficient to run systemd-based container on OpenShift.


@giuseppe may know more about the implementation, but it's possible that CRI-O and/or OpenShift automatically apply some configs and/or hooks to facilitate systemd inside a container.

There's a long history on systemd and containers with two camps (_"every container should have systemd as PID1"_ vs _"systemd doesn't belong in a container"_), which unfortunately led to a situation where either the user has to go through the (sometimes complicated) steps to make systemd work, or runtimes had to do preparations for this.

My personal opinion is that neither of those camps are "right", as it really depends on the use-case. In general, the default for _any_ container should be to have as little privileges as possible, and no _assumptions_ / _preparations_ should be made for a specific init system / tooling inside the container _by default_. _BUT_ I also think that it should be as easy as possible to be _able_ to use systemd (or other tools) in a container, and to be able to do so in a "portable" / "standard" way.

The OCI specifications are there to enhance portability; this is where I think "profiles" (or give it any other name) would be able to help. The OCI could define a "well known" profile for `systemd`, which would describe what preparations and/or changes to the container's spec runtimes are expected to make for containers using such profile (which could include permissions, environment variables, additional mounts and/or capabilities, etc.).

giuseppe replied:
I'd say this feature is needed more in general for enabling cgroup delegation when using a user namespace.

I don't think there should be any emphasis on systemd, since it is just an user of such feature.

For example on a cgroup v2 system:
```
$ podman --runtime /usr/bin/runc run --privileged --uidmap 0:1:1000 --rm fedora sh -c 'ls -lnd /sys/fs/cgroup; mkdir /sys/fs/cgroup/foo && echo created'
drwxr-xr-x. 2 65534 65534 0 Sep 13 08:02 /sys/fs/cgroup
mkdir: cannot create directory '/sys/fs/cgroup/foo': Permission denied
```

crun automatically chowns the cgroup to root in the userns:
```
$ podman --runtime /usr/bin/crun run --privileged --uidmap 0:1:1000 --rm fedora sh -c 'ls -lnd /sys/fs/cgroup; mkdir /sys/fs/cgroup/foo && echo created'
drwxr-xr-x. 2 0 0 0 Sep 13 08:02 /sys/fs/cgroup
created
```

frasertweedale replied:
@giuseppe indeed, systemd workload is just one notable use case.

Separately, your input over at https://github.com/opencontainers/runtime-spec/pull/1123 would be appreciated.

thaJeztah replied:
If this should be set unconditionally (and if it can be done without risk of introducing breaking changes), I'm definitely +1 on doing so; I agree that  https://github.com/opencontainers/runtime-spec/pull/1123 would be good for that (to define that in the spec).

AkihiroSuda replied:
If we are going to enable chowning by default, we will need to have a way to opt-out for that.

I guess skipping chowning when the mount is ro (https://github.com/opencontainers/runtime-spec/pull/1123#discussion_r707393997 , suggested by @giuseppe) would be fine.

frasertweedale replied:
I have updated the PR to remove the annotation, and only chown the cgroup when the cgroupfs is mounted `rw`.

frasertweedale commented:
OK, what is the next step for this PR.

I don't see any more objections to the overall implementation approach.
There is some debate/question over whether we need the annotation or not.

I also created a [runtime-spec PR](https://github.com/opencontainers/runtime-spec/pull/1123) to propose a formal semantics for cgroup ownership.
It looks like there will be a debate about whether to have an explicit config.  It may take
some time to resolve that matter.

Could we merge this PR in its current form, then remove or rename the annotation after
the runtime-spec discussion has been resolved?  Or, should we just remove the
annotation now, and try and convince runtime-spec to come to the party?  If there is a a doubt
let's be conservative and keep the annotation for now, so that this can be merged.

frasertweedale commented:
Could maintainers with authority to merge (I'm not sure who those are) please decide: do we remove the annotation now, or can we merge it as-is now, and potentially follow-up to remove the annotation later, as the runtime-spec outcome becomes clear?

This PR has dragged on a long time.  We need a clear decision from the maintainers so that it can be merged, even if that
decision is "we'll decide later whether to keep the annotation or make this behaviour unconditional".

ping @giuseppe @kolyshkin @AkihiroSuda @mrunalp

cyphar commented:
On paper I'd be okay with having an annotation, though unlike crun we've historically not really used annotations for extensions (mainly because runc is meant to be a reference implemetnation of the spec so out-of-spec stuff is in a bit of a grey area and extensions based on annotations could very quickly become an out-of-spec defacto standard like most other runc behaviour has become over the years). Unfortunately I don't see how we could add any features without this happening (and the horses have long-since bolted and ships long-since set sail on this topic -- so we're not going to solve it here).

As for making this behaviour the default, imho this entirely depends on whether it will break some existing programs that have faulty handling of cgroups. The other issue is that practically using annotations is somewhat difficult because image labels don't get translated to runtime labels by Docker AFAIK (podman/crio do though).

frasertweedale commented:
> On paper I'd be okay with having an annotation, though unlike crun we've historically not really used annotations for extensions (mainly because runc is meant to be a reference implementation of the spec so out-of-spec stuff is in a bit of a grey area and extensions based on annotations could very quickly become an out-of-spec defacto standard like most other runc behaviour has become over the years). Unfortunately I don't see how we could add any features without this happening (and the horses have long-since bolted and ships long-since set sail on this topic -- so we're not going to solve it here).
> 
> As for making this behaviour the default, imho this entirely depends on whether it will break some existing programs that have faulty handling of cgroups. The other issue is that practically using annotations is somewhat difficult because image labels don't get translated to runtime labels by Docker AFAIK (podman/crio do though).

@cyphar thanks for your feedback.  How to judge "whether it will break some existing programs"?  In the absence of a specific counterexample, it is an academic exercise in estimating the likelihood of breakage.  Or else, there is the "scream test" - merge it, see who screams, and be prepared to roll back or gate with an annotation if it's worse than anticipated.  Realistically, what do you think the risks of breakage are?

Personally I don't see much risk because this change strictly allows the container greater access on parts of its cgroupfs than it would otherwise have.  Container processes would have to be specifically inspecting the cgroupfs ownership or attempting/evaluating access to behave differently.  Or else, perhaps obscure UID maps could cause breakage (e.g. no uid 0 in the container's user namespace).

cyphar commented:
>  How to judge "whether it will break some existing programs"? In the absence of a specific counterexample, it is an academic exercise in estimating the likelihood of breakage. Or else, there is the "scream test" - merge it, see who screams, and be prepared to roll back or gate with an annotation if it's worse than anticipated. Realistically, what do you think the risks of breakage are?

This is a good question and aside from testing with a bunch of different programs (trying to get a representative sample) there isn't much more we can do than the scream test. But I think we should just merge this with the annotation-based flag and then when it is used by some downstreams we can see whether it caused any breakages. I hope it won't, and it seems somewhat unlikely but it's quite hard to predict what containerised programs try to do.

frasertweedale commented:
I pushed another update in response to https://github.com/opencontainers/runtime-spec/pull/1123#issuecomment-925464201.  Now the chown behaviour is also restricted to when runc will create a new cgroup namespace for the container.

frasertweedale commented:
In relation to @cyphar's recent comment - should I put the annotation guard back in?  I would like to hear from others on this - there seem to be people on both sides, within the last week.

frasertweedale commented:
I updated this PR to reflect the recent changes to the runtime-spec cgroup ownership
semantics proposed in PR https://github.com/opencontainers/runtime-spec/pull/1123. In particular:

- chown cgroup to host uid that maps to process.user.uid in container namespace (instead of uid 0 unconditionally)
- read list of files to chown from /sys/kernel/cgroup/delegate (with fallback if not exist)

I added these as separate commits to make them easier to review and, if necessary, modify or revert.
If desired, they can be squashed before this PR gets merged - just let me know.

Thanks, I look forward to your feedback.

jfroy commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -288,6 +290,36 @@ func (m *unifiedManager) Apply(pid int) error {
 		return err
 	}
+
+	// The kernel exposes a list of files that should chowned to
```
Perhaps move the code to compute `filesToChown` inside the `if c.OwnerUID != nil {` scope?

frasertweedale replied:
I shall implement this change in the next push.

jfroy commented on libcontainer/configs/cgroup_linux.go:
```diff
@@ -41,6 +41,13 @@ type Cgroup struct {
+	// cgroupfs is to be mounted read/write.
+	// Not all cgroup manager implementations support changing
+	// the ownership.
+	OwnerUID *int `json:"owner_uid,omitempty"`
```
Is there a need for group as well? crun sets both.

frasertweedale replied:
I'm not sure.  I don't need it for my use case.  It would be best to discuss it in the OCI runtime-spec PR (https://github.com/opencontainers/runtime-spec/pull/1123) where we are formalising a semantics for container cgroup ownership.

frasertweedale commented:
@jfroy thanks for your review.  I have implemented your proposed changes.  Discussion about whether group ownership should also be changed, it would be better to have that discussion over in https://github.com/opencontainers/runtime-spec/pull/1123.

frasertweedale commented:
The failing CI job (vagrant DISTRO:fedora) is failing when the commit tree is equal to that of master @ cbd725e6ef8a8c09ceea92b08d2530fc5af9a8ef.  See https://github.com/opencontainers/runc/runs/4149528511 for an
example.  Perhaps there is some defect in the CI environment?

The tests for the behaviour introduced by this PR are passing.  See https://cirrus-ci.com/task/5113759288000512:
```
...
ok 7 runc exec (cgroup v2, ro cgroupfs, new cgroupns) does not chown cgroup
ok 8 runc exec (cgroup v2, rw cgroupfs, inh cgroupns) does not chown cgroup
ok 9 runc exec (cgroup v2, rw cgroupfs, new cgroupns) does chown cgroup
...
```

AkihiroSuda commented:
The issues of Fedora CI are tracked in https://github.com/opencontainers/runc/issues/3266

frasertweedale commented:
Rebased, to absorb the CI fixes from main branch.

kolyshkin commented on libcontainer/specconv/spec_linux.go:
```diff
@@ -366,6 +366,39 @@ func CreateLibcontainerConfig(opts *CreateOpts) (*configs.Config, error) {
+	// namespace and the cgroupfs will be mounted read/write.
+	//
+	hasCgroupNS := config.Namespaces.Contains(configs.NEWCGROUP) && config.Namespaces.PathOf(configs.NEWCGROUP) == ""
+	hasRwCgroupfs := false
+	for _, m := range config.Mounts {
+		if m.Source == "cgroup" && m.Destination == "/sys/fs/cgroup" && (m.Flags&unix.MS_RDONLY) == 0 {
+			hasRwCgroupfs = true
+			break
+		}
+	}
```
In case `hasCgroupNS` is false, the results of this check is useless. This path can be quite common (for cgroup v1) so it probably makes sense to optimize it.

frasertweedale replied:
OK, I'll guard this loop behind `if hasCgroupNS`.

kolyshkin commented on tests/integration/cgroups.bats:
```diff
@@ -119,6 +119,86 @@ function setup() {
+	set_cgroups_path
+
+	# configure a user namespace
+	update_config '.linux.namespaces += [{"type": "user"}]'
+	update_config '.linux.uidMappings += [{"hostID": 100000, "containerID": 0, "size": 65536}]'
+	update_config '.linux.gidMappings += [{"hostID": 100000, "containerID": 0, "size": 65536}]'
```
A single update_config statement is enough. Not important, but we tend to use a single one elsewhere.

kolyshkin commented on tests/integration/cgroups.bats:
```diff
@@ -119,6 +119,86 @@ function setup() {
+	set_cgroup_mount_writable
+
+	# configure a user namespace
+	update_config '.linux.namespaces += [{"type": "user"}]'
+	update_config '.linux.uidMappings += [{"hostID": 100000, "containerID": 0, "size": 65536}]'
+	update_config '.linux.gidMappings += [{"hostID": 100000, "containerID": 0, "size": 65536}]'
```
ditto

kolyshkin commented on tests/integration/cgroups.bats:
```diff
@@ -119,6 +119,86 @@ function setup() {
+	update_config '.linux.uidMappings += [{"hostID": 100000, "containerID": 0, "size": 65536}]'
+	update_config '.linux.gidMappings += [{"hostID": 100000, "containerID": 0, "size": 65536}]'
+
+	# make cgroupfs mount read/write
+	update_config '.mounts |= map(select(.source == "cgroup").options |= map(select(. != "ro")) + ["rw"])'
```
Isn't it identical to `set_cgroup_mount_writable` above (and thus redundant)?

frasertweedale replied:
Thanks, it's vestigal, before I noticed `set_cgroup_mount_writable` was a thing.  Removing :)

kolyshkin commented on tests/integration/cgroups.bats:
```diff
@@ -119,6 +119,86 @@ function setup() {
+	update_config '.linux.uidMappings += [{"hostID": 100000, "containerID": 0, "size": 65536}]'
+	update_config '.linux.gidMappings += [{"hostID": 100000, "containerID": 0, "size": 65536}]'
+
+	# chown test temp dir to allow host user to read it
+	chown 100000 "$ROOT"
+
+	# chown rootfs to allow host user to mkdir mount points
+	chown 100000 "$ROOT"/bundle/rootfs
```
As you're doing this chowns and update_config in every test, maybe it make sense to move these to a separate file (e.g. `cgroup_delegate.bats`) and move the common part to `setup()` function.

frasertweedale replied:
Moving the tests to a separate file, sure.

Extracting the config manipulation to `setup()`, sure.

Unfortunately, these `chown`s can't be moved, because they need to be guarded by `requires root`.  Maybe that could also be moved to `setup()` but that feels wrong and likely to confuse.

frasertweedale replied:
But I did extract those `chown`s to a helper subroutine.

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -290,15 +291,27 @@ func (m *unifiedManager) Apply(pid int) error {
+		// file is not present (Linux < 4.15), use the initial
+		// values mentioned in in cgroups(7).
+		filesToChown := []string{"."} // the directory itself must be chowned
+		cgroupDelegateFile := "/sys/kernel/cgroup/delegate"
```
Ah! This is what I was talking about in https://github.com/opencontainers/runc/pull/3057/commits/56a353b13f93e60752ac52290df13e704db5d5d0#r700644360

kolyshkin replied:
So maybe merge this one into the first commit

frasertweedale replied:
Sure, I'm going to squash all commits.  Now that https://github.com/opencontainers/runtime-spec/pull/1123 was finally merged it is preferable to treat this whole PR as a single atomic commit.

kolyshkin reviewed (changes requested):
Mostly LGTM; I'd like for the last commit to be merged into the first one for clarity.

frasertweedale commented:
@kolyshkin thanks for your review.  I have implemented all of your suggestions.

frasertweedale commented:
ping @kolyshkin @AkihiroSuda @jfroy are there any further things to address for this PR?  Note that the runtime-spec PR https://github.com/opencontainers/runtime-spec/pull/1123 which proposed these cgroup ownership semantics has been merged.

AkihiroSuda commented on libcontainer/specconv/spec_linux.go:
```diff
@@ -366,6 +366,49 @@ func CreateLibcontainerConfig(opts *CreateOpts) (*configs.Config, error) {
+	hasRwCgroupfs := false
+	if hasCgroupNS {
+		for _, m := range config.Mounts {
+			if m.Source == "cgroup" && m.Destination == "/sys/fs/cgroup" && (m.Flags&unix.MS_RDONLY) == 0 {
```
```suggestion
			if m.Source == "cgroup" && filepath.Clean(m.Destination) == "/sys/fs/cgroup" && (m.Flags&unix.MS_RDONLY) == 0 {
```

frasertweedale replied:
@AkihiroSuda thank you.  I updated to the PR to implement this change, which you requested.

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -288,6 +290,36 @@ func (m *unifiedManager) Apply(pid int) error {
+		// The kernel exposes a list of files that should chowned to
+		// the delegate uid in /sys/kernel/cgroup/delegate.  If the
+		// file is not present (Linux < 4.15), use the initial
+		// values mentioned in in cgroups(7).
```
typo: `in in`.

frasertweedale replied:
thanks!  fixed in the latest push.

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -288,6 +290,36 @@ func (m *unifiedManager) Apply(pid int) error {
 	}
+
+	if c.OwnerUID != nil {
+		// The kernel exposes a list of files that should chowned to
```
typo: should **be** chowned.

frasertweedale replied:
thanks!  fixed in the latest push.

frasertweedale commented on tests/integration/cgroup_delegation.bats:
```diff
@@ -0,0 +1,76 @@
+}
+
+@test "runc exec (cgroup v2, ro cgroupfs, new cgroupns) does not chown cgroup" {
+	requires root cgroups_v2 systemd
```
done, thanks.

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -288,6 +290,36 @@ func (m *unifiedManager) Apply(pid int) error {
+			filesToChown = append(filesToChown, "cgroup.procs", "cgroup.subtree_control", "cgroup.threads")
+		}
+
+		for _, v := range filesToChown {
```
Perhaps it's better to separate out the code that finds files to be chowned to a separate `filesToChown()` function, and just call it from below, e.g.

```go
for _, v := range filesToChown() {
...
}
```

This should increase readability and decrease Apply() complexity.

The only thing is, to simplify, you would not return any errors from filesToChown(), instead maybe logging it as warnings and falling back to built-in list.

frasertweedale replied:
I extracted the behaviour to another function.  But I kept the error return.  I feel that `/sys/kernel/cgroup/delegate` existing but being unreadable is something we should fail hard on, because its values may (and currently do) differ from the original list.

kolyshkin replied:
Thanks, this looks better and more reviewable!

kolyshkin reviewed (changes requested):
Overall looks good and ready; only left a single suggestion.

kolyshkin commented on libcontainer/cgroups/systemd/v2.go:
```diff
@@ -288,9 +290,46 @@ func (m *unifiedManager) Apply(pid int) error {
+// (Linux < 4.15), use the initial values mentioned in cgroups(7).
+func cgroupFilesToChown() ([]string, error) {
+	filesToChown := []string{"."} // the directory itself must be chowned
+	cgroupDelegateFile := "/sys/kernel/cgroup/delegate"
```
Nit: this can be `const`.

frasertweedale replied:
Thanks! Done.

kolyshkin reviewed (approved):
LGTM

kolyshkin commented on libcontainer/specconv/spec_linux.go:
```diff
@@ -366,6 +366,49 @@ func CreateLibcontainerConfig(opts *CreateOpts) (*configs.Config, error) {
+	hasRwCgroupfs := false
+	if hasCgroupNS {
+		for _, m := range config.Mounts {
+			if m.Source == "cgroup" && filepath.Clean(m.Destination) == "/sys/fs/cgroup" && (m.Flags&unix.MS_RDONLY) == 0 {
```
Ideally, the simplest check (m.Flags is RO) should be the first one, to make things slightly faster, but it's not a big deal, plus the way it's written is more readable.

frasertweedale replied:
Yes, I think we can spend a few CPU cycles to save some brain cycles :)

kolyshkin reviewed (approved):
LGTM

frasertweedale commented:
Thanks to all who have reviewed this PR and given feedback.  We need more approvals and then someone can press the merge button.  Ping @cyphar, @AkihiroSuda, @giuseppe, @jfroy.

frasertweedale commented:
Any takers?  @mrunalp?

cyphar reviewed (approved):
LGTM.

frasertweedale commented:
Many thanks, @cyphar and all reviewers.

frasertweedale commented:
I created https://github.com/opencontainers/runc/pull/3311 to backport this feature to the `release-1.0` branch.
